doc/developer/design/20230421_stabilize_monotonic_select.md

```diff
@@ -0,0 +1,137 @@
+
+Another alternative to the above proposal is to not retire the feature flag `enable_monotonic_oneshot_selects` and run selected tests from existing corpora with the flag either turned on or off. While simpler to implement, this solution goes against the notion that feature flags should eventually meet an endpoint in their lifecycle.
+
+## Lowering the Expense of Forced Consolidation
+
+There are two strategies that we propose to lower the expense of monotonic operators introduced with forced consolidation during single-time LIR `Plan` refinement.
+
+### Improve `refine_single_time_dataflow` LIR-based Refinement (MaterializeInc/materialize#18732)
+
+It is not always the case the `must_consolidate` must be set for every operator coverted to monotonic in a one-shot `SELECT`. For example, in a single-time dataflow, some operators may produce physically monotonic output regardless of whether their input was even logically monotonic, e.g., reductions. Furthermore, some operators will convert single-time, logically monotonic input into physically monotonic input, e.g., operators creating an arrangement. So if the input to a monotonic operator with forced consolidation came from a reduction, then this operator's `must_consolidate` flag could instead be toggled to `false`.
+
+We propose to split the `refine_single_time_dataflow` LIR-based refinement into two sub-actions. The first consists of the present code, which replaces non-monotonic operators with monotonic ones where consolidation is enforced (e.g., `refine_single_time_operator_selection`). The second takes as input a collection of references for the monotonic operators that were replaced in the plan and refines the setting of the `must_consolidate` flag for these `Plan` nodes (e.g., `refine_single_time_consolidation`).
+
+The `must_consolidate` flag refinement can be implemented as a series of iterative child traversals starting from each monotonic `Plan` node replaced in the first sub-action. For a given monotonic `Plan` node, we would like to assert whether its input is already physically monotonic or whether consolidation must be forced in the `Plan` node itself. To support this decision in a single-time context, we note that `Plan` nodes can be classified into: (a) Monotonicity markers, e.g., `Plan::Reduce` or `Plan::Join`, which are LIR nodes that require either arranged input or enforce arranged output; (b) Monotonicity keepers, e.g., `Plan::FlatMap`, which just maintain the monotonicity property of their input unchanged; or (c) Monotonicity breakers, e.g., `Plan::Negate` or `Plan::Get` on a non-arranged and non-monotonic source, which introduce negative multiplicities, thus destroying physical monotonicity.
```

Comment:
> (a) Monotonicity markers

Nit: I stumbled a bit over the term "markers" here. Considering that we have "keepers" and "breakers", I'd call these "makers" instead.

Reply:
I was a bit in doubt about the terminology here too. I chose "markers" instead of "enforcers" (similar to "makers"), because a `Plan::Join`, for example, is not really converting logically monotonic input to physically monotonic input. It just _requires_ that its inputs be so converted by forcing them to be arranged. 

So the "enforcers"/"makers" are actually the underlying arrange operators; but we can already infer physical monotonicity about the output of the join due to the constraints that the join operator puts on its inputs. Not sure there would be a term that is better than "marker" here, but that would capture the idea that this is not the operator actually enforcing/making the physical monotonicity property be true?

Reply:
That makes sense. I think "marker" is fine in this case. Alternatively, how about using "enforcers/makers" and instead excluding `Join` from that category? That would not affect correctness, since our analysis would ignore the join and instead find the arrange operator (right?) and it would allow us to have one special case less. I'm not sure about the implications on the search performance though.